---
title: "RNA-seq and transcriptomics"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Read our data for today

```{r}
counts <- read.csv("airway_scaledcounts.csv", stringsAsFactors = FALSE)
metadata <- read.csv("airway_metadata.csv", stringsAsFactors = FALSE)
```

How many genes are we dealing with?

```{r}
nrow(counts)
```

Now we want to just compare, take the average of all the control rows and all the treated rows. 
First we will extract the control columns by looking up their colnames in the `metadata` object

```{r}
metadata
```

We see the dex column has the group info we need, and the ID has the corresponding counts, so can we extract all the control rows and get their average

```{r}
metadata[metadata$dex=="control",] #gets all rows that are part of the control group: dex has the distinction, we want only control ROWS
metadata[metadata$dex=="control",]$id #gets us the name of these columns
```


```{r}
control <- metadata[metadata$dex=="control",]
control
```

Now access the count columns 

```{r}
counts[, control$id] #we've created a path to get all the control rows, which correspond to the COLUMNS in the actual counts data set
```

Now find the mean value for each row (aka each gene)

```{r}
control.mean <- rowSums(counts[, control$id]) / nrow(control) # aka 4 but you want to make this applicable to other data sets
control.mean
```

Now access the treated rows and make a new vector with their means

```{r}
treated <- metadata[metadata$dex=="treated",]
treated
counts[, treated$id]
treated.mean <- rowSums(counts[, treated$id]) / nrow(treated)
treated.mean
```

Lets put these together for ease of use

```{r}
mycounts <- data.frame(control.mean, treated.mean)
head(mycounts)
```

Lets plot and see what it looks like

```{r}
plot(mycounts)
```

```{r}
hist(mycounts$control.mean, breaks = 300)
```

Whenever you see something like this TAKE THE LOG

```{r}
plot(mycounts, log = "xy") # determines which axis you want to be on the log scale
```

Now we'll try to remove all the zero count genes because we cant say anything about them from this dataset/theyre messing up our plot

```{r}
x <- c(1,3,10,0) #demo
which(x == 0)

y <- data.frame( c(1,3,10,0),
                 c(1,2,0,0))

unique(which(y == 0, arr.ind = TRUE) [, "row"] ) 
#this gives us the row and col where the zeros are, more details
#gives us back unique entries

```

Now actually find zero rows in our mycounts object

```{r}
to.rm <- unique(which(mycounts == 0, arr.ind = TRUE)[, "row"])
newcounts <- mycounts[-to.rm,] # gives you all the data without the zeros
head(newcounts)
nrow(newcounts)
```

Now we have 21,817 genes, which is a little more reasonable given what we know about the number of existing genes, remember to take the log

```{r}
newcounts$log2fc <- log2(newcounts[,"treated.mean"]/newcounts[,"control.mean"])
head(newcounts)
```

Now lets look at the genes that actually changed

```{r}
upreg.genes <- newcounts$log2fc > 2 #bc we used log 2
sum(upreg.genes)
```

```{r}
downreg.genes <- newcounts$log2fc < (-2) 
sum(downreg.genes)
```

## Lets use DESeq2

```{r}
library(DESeq2)
```

Set up of the object required by DESeq

```{r}
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex, 
                              tidy=TRUE)
dds
```

```{r}
dds <- DESeq(dds)
```

```{r}
res <- results(dds)
res
```

Time to visualize the data! Volcano Plot - plot of fold change vs adjusted pvalue

```{r}
plot(res$log2FoldChange, - log(res$padj),
          xlab="Log2(FoldChange)",
         ylab="-Log(P-value)", col = "grey")

# Setup our custom point color vector 
mycols <- rep("gray", nrow(res))
mycols[ abs(res$log2FoldChange) > 2 ]  <- "red" 

inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

# Volcano plot with custom colors 
plot( res$log2FoldChange,  -log(res$padj), 
 col=mycols, ylab="-Log(P-value)", xlab="Log2(FoldChange)" )

#setting up boundaries
abline(v=c(-2,+2), lty=2) 
abline(h=c(-log(0.05)), lty = 2)

```
















